File: Python/BFS/BFS (adjacency list).py
from collections import deque


# adjacency list
def adj_check(node1, node2):
    global adj

    if node1 not in adj.keys():
        adj[node1] = []

    adj[node1].append(node2)


def bfs(num):
    global adj, q, ans, visited

    q.append(num)

    while len(q) > 0:
        cur = q.popleft()
        
        if not visited[cur]:
            ans.append(cur)
        
        visited[cur] = True

        if cur not in adj.keys():
            continue

        for elem in adj[cur]:
            if not visited[elem]:
                q.append(elem)


def ans_print():
    global ans
    print("bfs search order")

    while len(ans) > 0:
        print("->{}".format(ans.popleft()), end="")


if __name__ == '__main__':
    visited = [False for _ in range(15)]
    adj = {}
    q = deque()
    ans = deque()

    # 1: [2,6,7,9]
    adj_check(1, 2)
    adj_check(1, 6)
    adj_check(1, 7)
    adj_check(1, 9)

    # 2: [3, 4]
    adj_check(2, 3)
    adj_check(2, 4)

    # 4: [5]
    adj_check(4, 5)

    # 6: [4]
    adj_check(6, 4)

    # 7:[8]
    adj_check(7, 8)

    bfs(1)
    ans_print()

File: C++/BFS/BFS (adjacency list).cpp
#include<iostream>
#include<queue>
#include<vector>
using namespace std;

bool visited[15];
vector <int> adj[15];

queue<int> q;
queue<int> ans;

void adj_check()
{
    // adj[1][2,6,7,9]
    adj[1].push_back(2);
    adj[1].push_back(6);
    adj[1].push_back(7);
    adj[1].push_back(9);
    
    //adj[2][3,4]
    adj[2].push_back(3);
    adj[2].push_back(4);
    
    //adj[4][5]
    adj[4].push_back(5);
    
    //adj[7][8]
    adj[7].push_back(8);
}

void bfs(int num)
{
    visited[num] = true;
    q.push(num);
    
    while(!q.empty())
    {
        int cur = q.front(); q.pop();
        visited[cur] = true;
        ans.push(cur);
        
        for(auto elem: adj[cur])
        {
            if(!visited[elem])
                q.push(elem);
        }
        
    }
}

void ans_print()
{
    cout << "bfs search order\n";
    cout << ans.front(); ans.pop();
    while(!ans.empty())
    {
        int cur = ans.front(); ans.pop();
        cout << "->"<<cur;
    }
    cout <<"\n";
}

int main()
{
    adj_check();
    bfs(1);
    ans_print();
    exit(0);
}


